Build failure: graalvm-ce on x86_64-darwin

### Steps To Reproduce

1. build graalvm-ce on x86_64-darwin

### Build log

https://hydra.nixos.org/build/276756639

### Notify maintainers

@bennyandresen @hlolli @glittershark @ericdallo @thiagokokada 

### Metadata

Build was attempted on Hydra.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

> Looks like a regression? #299203

This doesn't look like to be the same issue. I think the reason is the recent macOS SDK bump.

``` 
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libnio.a(UnixFileSystem.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libnio.a(UnixNativeDispatcher.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(TimeZone_md.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(UnixFileSystem_md.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libnio.a(nio_util.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(VM.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(VMSupport.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(VM_md.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(VirtualThread.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(canonicalize_md.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(check_classname.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(check_version.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(childproc.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(io_util.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(io_util_md.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(java_props_macosx.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(java_props_md.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(jio.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(jni_util.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(jni_util_md.o)) was built for newer macOS version (11.0) than being linked (10.12)
ld: warning: object file (/nix/store/75r6jjjlb65ikiwjjw550p6app04jacq-graalvm-ce-23.0.0/lib/static/darwin-amd64/libjava.a(path_util.o)) was built for newer macOS version (11.0) than being linked (10.12)
```

Unfortunaly I can't figure out the issue here. The obvious fix would be to set `MACOSX_DEPLOYMENT_TARGET` to `11.0` (we also need to add to the wrapper `native-image` wrapper), but it results in another error:

```
Undefined symbols for architecture x86_64:
  "____chkstk_darwin", referenced from:
      _InflateFully in libzip.a(zip_util.o)
      _checkCPUFeatures in liblibchelper.a(cpuid.o)
      _readBytes in libjava.a(io_util.o)
      _writeBytes in libjava.a(io_util.o)
      _collapse in libjava.a(path_util.o)
ld: symbol(s) not found for architecture x86_64
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.NativeImageViaCC.handleLinkerFailure(NativeImageViaCC.java:204)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.NativeImageViaCC.runLinkerCommand(NativeImageViaCC.java:154)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:115)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:761)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:554)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:528)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:711)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.start(NativeImageGeneratorRunner.java:139)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:94)
```

--- a/pkgs/build-support/build-graalvm-native-image/default.nix
+++ b/pkgs/build-support/build-graalvm-native-image/default.nix
@@ -1,6 +1,8 @@
 {
   lib,
   stdenv,
+  apple-sdk_11,
+  darwinMinVersionHook,
   glibcLocales,
   # The GraalVM derivation to use
   graalvmDrv,
@@ -33,6 +35,8 @@ let
   extraArgs = builtins.removeAttrs args [
     "lib"
     "stdenv"
+    "apple-sdk_11"
+    "darwinMinVersionHook"
     "glibcLocales"
     "jar"
     "dontUnpack"
@@ -54,6 +58,11 @@ stdenv.mkDerivation (
       graalvmDrv
       glibcLocales
       removeReferencesTo
+    ];
+
+    buildInputs = lib.optionals (stdenv.hostPlatform.isDarwin) [
+      apple-sdk_11
+      (darwinMinVersionHook "11.0")
     ];
 
     nativeImageBuildArgs = nativeImageBuildArgs ++ extraNativeImageBuildArgs ++ [ graalvmXmx ];

--- a/pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix
+++ b/pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix
@@ -2,10 +2,11 @@
   lib,
   stdenv,
   alsa-lib,
+  apple-sdk_11,
   autoPatchelfHook,
   cairo,
   cups,
-  darwin,
+  darwinMinVersionHook,
   fontconfig,
   glib,
   glibc,
@@ -30,10 +31,12 @@ let
     "lib"
     "stdenv"
     "alsa-lib"
+    "apple-sdk_11"
     "autoPatchelfHook"
     "cairo"
     "cups"
     "darwin"
+    "darwinMinVersionHook"
     "fontconfig"
     "glib"
     "glibc"
@@ -122,34 +125,51 @@ let
       propagatedBuildInputs = [
         setJavaClassPath
         zlib
-      ] ++ lib.optional stdenv.hostPlatform.isDarwin darwin.apple_sdk_11_0.frameworks.Foundation;
-
-      buildInputs = lib.optionals stdenv.hostPlatform.isLinux [
-        alsa-lib # libasound.so wanted by lib/libjsound.so
-        fontconfig
-        (lib.getLib stdenv.cc.cc) # libstdc++.so.6
-        xorg.libX11
-        xorg.libXext
-        xorg.libXi
-        xorg.libXrender
-        xorg.libXtst
       ];
+
+      buildInputs =
+        lib.optionals stdenv.hostPlatform.isLinux [
+          alsa-lib # libasound.so wanted by lib/libjsound.so
+          fontconfig
+          (lib.getLib stdenv.cc.cc) # libstdc++.so.6
+          xorg.libX11
+          xorg.libXext
+          xorg.libXi
+          xorg.libXrender
+          xorg.libXtst
+        ]
+        ++ (lib.optionals stdenv.hostPlatform.isDarwin [
+          apple-sdk_11
+          (darwinMinVersionHook "11.0")
+        ]);
 
       postInstall =
         let
           cLibsAsFlags = (map (l: "--add-flags '-H:CLibraryPath=${l}/lib'") cLibs);
-          preservedNixVariables = [
-            "-ELOCALE_ARCHIVE"
-            "-ENIX_BINTOOLS"
-            "-ENIX_BINTOOLS_WRAPPER_TARGET_HOST_${stdenv.cc.suffixSalt}"
-            "-ENIX_BUILD_CORES"
-            "-ENIX_BUILD_TOP"
-            "-ENIX_CC"
-            "-ENIX_CC_WRAPPER_TARGET_HOST_${stdenv.cc.suffixSalt}"
-            "-ENIX_CFLAGS_COMPILE"
-            "-ENIX_HARDENING_ENABLE"
-            "-ENIX_LDFLAGS"
-          ];
+          preservedNixVariables =
+            [
+              "-ENIX_BINTOOLS"
+              "-ENIX_BINTOOLS_WRAPPER_TARGET_HOST_${stdenv.cc.suffixSalt}"
+              "-ENIX_BUILD_CORES"
+              "-ENIX_BUILD_TOP"
+              "-ENIX_CC"
+              "-ENIX_CC_WRAPPER_TARGET_HOST_${stdenv.cc.suffixSalt}"
+              "-ENIX_CFLAGS_COMPILE"
+              "-ENIX_HARDENING_ENABLE"
+              "-ENIX_LDFLAGS"
+            ]
+            ++ lib.optionals stdenv.hostPlatform.isLinux [
+              "-ELOCALE_ARCHIVE"
+            ]
+            ++ lib.optionals stdenv.hostPlatform.isDarwin [
+              "-EDEVELOPER_DIR"
+              "-EDEVELOPER_DIR_FOR_BUILD"
+              "-EDEVELOPER_DIR_FOR_TARGET"
+              "-EMACOSX_DEPLOYMENT_TARGET"
+              "-EMACOSX_DEPLOYMENT_TARGET_FOR_BUILD"
+              "-EMACOSX_DEPLOYMENT_TARGET_FOR_TARGET"
+              "-ENIX_APPLE_SDK_VERSION"
+            ];
           preservedNixVariablesAsFlags = (map (f: "--add-flags '${f}'") preservedNixVariables);
         in
         ''

--- a/pkgs/development/compilers/graalvm/community-edition/default.nix
+++ b/pkgs/development/compilers/graalvm/community-edition/default.nix
@@ -4,9 +4,6 @@
 }:
 
 lib.makeScope pkgs.newScope (self: {
-  stdenv =
-    if pkgs.stdenv.hostPlatform.isDarwin then pkgs.darwin.apple_sdk_11_0.stdenv else pkgs.stdenv;
-
   buildGraalvm = self.callPackage ./buildGraalvm.nix;
 
   buildGraalvmProduct = self.callPackage ./buildGraalvmProduct.nix;

